src/folding/nova/ivc.rs

```diff
@@ -82,67 +119,87 @@ where
             u_i: u_dummy.clone(),
             W_i: w_dummy,
             U_i: u_dummy,
+            // cyclefold running instance
+            cf_W_i: cf_w_dummy.clone(),
+            cf_U_i: cf_u_dummy.clone(),
         })
     }
 
     /// Implements IVC.P
     pub fn prove_step(&mut self) -> Result<(), Error> {
         let u_i1_x: C1::ScalarField;
-        let augmented_F_circuit: AugmentedFCircuit<C1, FC>;
+
+        let augmented_F_circuit: AugmentedFCircuit<C1, C2, GC2, FC>;
+        let cf_circuit: CycleFoldCircuit<C1, GC1>;
+
         let z_i1 = self.F.step_native(self.z_i.clone());
 
-        let (W_i1, U_i1, cmT): (Witness<C1>, CommittedInstance<C1>, C1);
+        let (W_i1, U_i1): (Witness<C1>, CommittedInstance<C1>);
 
         if self.i == C1::ScalarField::zero() {
-            // base case: i=0, z_i=z_0, U_i = U_d := dummy instance
-            // u_1.x = H(1, z_0, z_i, U_i)
-            u_i1_x = self.U_i.hash(
+            let (T, cmT) = self.compute_cmT()?;
+
+            let r_bits = ChallengeGadget::<C1>::get_challenge_native(
+                &self.poseidon_config,
+                self.u_i.clone(),
+                self.U_i.clone(),
+                cmT,
+            )?;
+            let r_Fr = C1::ScalarField::from_bigint(BigInteger::from_bits_le(&r_bits))
+                .ok_or(Error::OutOfBounds)?;
+
+            // fold dummy (initial) instances, compute W_{i+1} and U_{i+1}
+            (W_i1, U_i1) = NIFS::<C1>::fold_instances(
+                r_Fr, &self.w_i, &self.u_i, &self.W_i, &self.U_i, &T, cmT,
+            )?;
+
+            // u_i+1.x = H(1, z_0, z_i+1, U_1=U_i+1)
+            u_i1_x = U_i1.hash(
```

Comment:
ditto, these parts are same with non-base case, we could move these lines out of if-else and save some lines of code.

Reply:
Oh right! Updated in https://github.com/privacy-scaling-explorations/folding-schemes/pull/45/commits/194923d09f5b79e180855b94a7aec3047cfa91bf